### ISSUE TITLE
Can't build using QT 5.15

### ISSUE THREAD

**Issue by lemmy04** [Stellarium/stellarium#1131]
I'm packaging stellarium for openSUSE, and on the openSUSE build service Qt has been upgraded to 5.15 for most build targets that I'm covering.
Since this upgrade stellarium fails to build.

### Actual Behaviour
you can find the build log here:
https://build.opensuse.org/package/live_build_log/home:lemmy04/stellarium/openSUSE_Leap_15.1/x86_64

### Steps to reproduce
Build using QT 5.15

### System
* Stellarium version: 0.20.1
* Operating system: building from source for openSUSE Leap

**Comment by gzotti** [comment]
Yes, I also have problems, could build on one Win10, but not on the other. Something in the translations failed.

**Comment by alex-w** [comment]
Cannot reproduce on Manjaro with Qt 5.15

**Comment by gzotti** [comment]
I see one probably decisive hint:

`[  149s] [  138.097088] Out of memory: Kill process 3703 (lconvert) score 179 or sacrifice child`

Must see if that also is the "killer" in my one case. Any method to mitigate this? Force make to run less jobs? Or how can we push lconvert to OOM in the first place?

**Comment by arjendekorte** [comment]
Fixed in https://build.opensuse.org/request/show/816443

**Comment by gzotti** [comment]
@arjendekorte I see you configured make -j1. This is a workaround which can hopefully be improved. Any idea why lconvert is suddenly freaking out with memory consumption?

**Comment by arjendekorte** [comment]
I noticed the failing builds about a week ago. It seems the memory requirements have steadily increased with every major  upgrade of Qt. In Qt 5.6, the peak memory was just over 1 GB, in the latest Qt 5.15 it is almost 4 GB.

The reason we see it failing now, is probably that the default memory limit for the buildservers is now exceeded (a single lconvert process can now take almost 2 GB of memory). If two of them are running in parallel, most (if not all) buildservers will run OOM.

In the buildservers we might work around this by using a _constraints file to require a certain amount of memory and limit the number of cores used, but this limits the number of eligible buildservers and usually will cause delayed building. This makes no sense here.

I'm afraid the only thing we can do is to prevent multiple lconvert processes running simultaneously. There may be other ways to achieve this, though.

**Comment by alex-w** [comment]
Tried build Stellarium with Qt 5.15 and MSVC 2019 on Windows 7 x64 at work - no problem

**Comment by arjendekorte** [comment]
@alex-w Does this build in parallel (/MP) by default? If so, how many cores are used for build? What I see happening is memory exhaustion if a sufficient number of lconvert processes is running in parallel.

**Comment by alex-w** [comment]
2 core was  used

**Comment by arjendekorte** [comment]
@alex-w Any idea about maximum memory consumption?

**Comment by gzotti** [comment]
I have updated the Qt 5.15 packages today to the latest versions. Now it seems I could build on both (secondary) systems, with 8GB (Core-i5, 4 cores) and 12GB (Core-i7, 8 threads) RAM. What was weird is that the 12GB system had failed, presumably running OOM with more worker threads.  
Here is the memory consumption on the 8GB machine. While running translations, there is a weird sawtooth pattern (apparently overlaying single translation jobs) where memory requirements go up and down by ~3GB. The lower memory consumption values start when regular code compilation starts.
![Stel_Qt5 15-win64-mem3](https://user-images.githubusercontent.com/18099873/87768039-d4516e00-c81b-11ea-8330-2594f70ef898.png)

I really don't know if that is intended behaviour, a bug in lconvert, or if we are doing something wrong regarding the application of the tools. However, for memory-depraved systems (in today's standards, 2-6GB?) there should be an easily accessible switch in QtCreator settings to reduce building (if possible, only application of lconvert) to 1-2 threads. I am however unable to find such switch. (I have a 4GB field netbook that so far "works" with MSVC2017/Qt5.12...)

**Comment by arjendekorte** [comment]
Citeren Georg Zotti <notifications@github.com>:

> I have updated the Qt 5.15 packages today to the latest versions.  
> Now it seems I could build on both (secondary) systems, with 8GB  
> (Core-i5, 4 cores) and 12GB (Core-i7, 8 threads) RAM. What was weird  
> is that the 12GB system had failed, presumably running OOM with more  
> worker threads.
> Here is the memory consumption on the 8GB machine. While running  
> translations, there is a weird sawtooth pattern (apparently  
> overlaying single translation jobs) where memory requirements go up  
> and down by ~3GB. The lower memory consumption values start when  
> regular code compilation starts.
> ![Stel_Qt5  
> 15-win64-mem3](https://user-images.githubusercontent.com/18099873/87768039-d4516e00-c81b-11ea-8330-2594f70ef898.png)
>
> I really don't know if that is intended behaviour, a bug in  
> lconvert, or if we are doing something wrong regarding the  
> application of the tools. However, for memory-depraved systems (in  
> today's standards, 2-6GB?) there should be an easily accessible  
> switch in QtCreator settings to reduce building (if possible, only  
> application of lconvert) to 1-2 threads. I am however unable to find  
> such switch. (I have a 4GB field netbook that so far "works" with  
> MSVC2017/Qt5.12...)

In the openSUSE buildservice, we pass the -j1 option to the make  
command to limit building to one core only if Qt >= 5.15. There is  
little benefit in building on multiple cores (the builds are scheduled  
and run without human intervention) and we found that indeed the  
memory requirement per core to be in the 2+ GB range when running  
lconvert. Sadly there is no way to allow a variable number of cores  
and still guarantee at least 2 GB of memory per core in the  
buildservice, so we opted to limit building to one core only.

**Comment by gzotti** [comment]
Yes, that's OK-ish [workaround] for unattended nightly builds, but terrible while developing. (And I just don't find where to configure the equivalent of -j1 on Windows!)

**Comment by arjendekorte** [comment]
You can limit the number of cores used for building with the /MP<n> option (where <n> is a number from 1 to 65535). See https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2019

**Comment by arjendekorte** [comment]
For development (where you always know your configuration and set <n> appropriately to get at least 3 GB per core) this might work, but there doesn't seem to be an option to determine this automatically on Windows either.

**Comment by gzotti** [comment]
The /MP is only for the compiler. This does not cause any problem. lconvert is the bad guy.

However, I have now updated the 4GB netbook (1 GHz/dual core) to MSVC2019+Qt5.15.0. With this current edition, I was able to build Stellarium. (>3hours...) 

Probably the tools update from Qt fixed the problem. Next I dare to upgrade my main system.

**Comment by alex-w** [comment]
I see no problem with lconvert on Win7 (MSVC2019+Qt5.15.0) with 4GB RAM, but you see problem with it on Win10/12GB RAM - changes in OOM?

**Comment by gzotti** [comment]
I have no idea how lconvert works internally, and why apparently one lconvert process (or some other subtask called during translation, xgettext, python?) takes considerably more than 1GB of RAM to process po files just over 2MB in size (planetary features). In May/June, this failed on a 12GB/4 core/8 thread notebook, but worked on an 8GB/4core/4thread PC. The RAM pattern (above) is definitely crazy.  
I have not found in QtCreator's menu how or where to limit the number of parallel tasks on Windows. (I am well aware of "make -jN" on Linux. Qt uses other tools on Windows. Ninja? Jom? Sorry, had no time for longer search.) The actual compilation of C++ code takes so little memory that parallel execution of the compiler should not be prevented.
 
As of mid-July, after Qt online update of whatever was available, I can build now on both computers, and even on my 4GB/2 core/2 thread netbook. But still, as you see, the RAM consumption of lconvert is enormous. I have no 16GB/12core Ryzen (or similar) to test, but I would not be surprised if that failed again. 

If your Win7 is in a VM, can you try to trigger OOM by building with 4 or 8 cores?

**Comment by alex-w** [comment]
This is real hardware (my work PC)

**Comment by gzotti** [comment]
Can you observe similar memory peaks in task manager while building? Maybe I need to install/update some better external tools (which cmake may find)?

**Comment by alex-w** [comment]
lconvert used up to 3GB RAM per process on huge planetary-features files (I'm use task manager and resource monitor) and at first run I got the error (jom/ninja), on the second start of run it was works.

**Comment by gzotti** [comment]
OK. Whatever I did in May, it failed consistently. Now after update it works. But do you think 3GB for processing 2.5MB .po's (or which files?) is "normal behaviour"? What happens here? And why won't it just allocate virtual memory? (Slow but so what...)

**Comment by alex-w** [comment]
My OS allocated virtual memory for lconvert... ;)

**Comment by alex-w** [comment]
Please check Qt 5.15.1

**Comment by arjendekorte** [comment]
No joy with parallel builds, the memory requirements of lconvert are still excessive:

[   70s] [   63.355884] Out of memory: Killed process 2782 (lconvert) total-vm:2516920kB, anon-rss:1393148kB, file-rss:4kB, shmem-rss:0kB, UID:399 pgtables:4924kB oom_score_adj:0

**Comment by alex-w** [comment]
Bad news :(

Theoretically we can try define limits for RAM in cmake/ninja, but this is not very good.

**Comment by gzotti** [comment]
On my first updated Win10 PC (Core i5-4xxx), I see up to 4.6GB memory consumption during translation building, then about 550MB while compiling with all kernels. This is still crazy, but works. Need to update a few more systems and test, though... 

It works then OK, with the usual exchange of libEGL.dll and libGLESv2.dll from older Qt versions.

**Comment by arjendekorte** [comment]
> Bad news :(
> 
> Theoretically we can try define limits for RAM in cmake/ninja, but this is not very good.

This would probably mean that we wouldn't be able to build the translations anymore. I'm still perplexed why a lconvert process would need more than 2.5 GB of memory. That's 2.5 times the latest edition of the entire Encyclopaedia Britannica (all 32 volumes). Setting limits or upping memory requirements doesn't seem to be the solution here.

**Comment by gzotti** [comment]
Updated my second system (Core i7-2xxx, 12GB RAM).  Now builds also with 5.15.1, running 8 parallel translation jobs. At least pagefile allocation now works, and it does not crash :-|. However, this translation memory issue causes considerable slowdown now compared to build times I remember a year ago. The greatest consumption is during translation of planetary feature names, which is for most languages presumably a 1->1 copy (unless a transliteration to local character sets is done). Memory consumption during code compiling  is below 1GB.
I know that an untranslated string is just taken from the English one. To keep the package smaller by several MBs, would it be possible to specify "Don't apply translations, and don't offer translations on Transifex", on a per-language setting? Or would it be useful to split the planetary features e.g. per-planet? (If allocation grows so rapidly with number of strings?)

Maybe we should report a Qtbug after all, but I'd need to know more about how our translation really works. Qt may assume a different workflow. I have no idea what lconvert is doing internally with GBs of memory.

**Comment by gzotti** [comment]
I filed https://bugreports.qt.io/browse/QTBUG-87010

**Comment by lemmy04** [comment]
FYI: I'm restricting the CI environment to "make -j1" now, and to a build VM with 4GB Ram, and it seems to build ok now - but it takes ages.

**Comment by arjendekorte** [comment]
> FYI: I'm restricting the CI environment to "make -j1" now, and to a build VM with 4GB Ram, and it seems to build ok now - but it takes ages.

This will work for the x86_64 targets, but so de the latest version 0.20.3 will fail to build on i586.

**Comment by lemmy04** [comment]
good point. Didn't think of that because I'm building for openSUSE Leap 15.2 which only comes in X86_64.

**Comment by danfe** [comment]
Version 0.20.3 also does not build on FreeBSD/i386 against Qt 5.15 (while 0.20.2 apparently did).

**Comment by arjendekorte** [comment]
> Version 0.20.3 also does not build on FreeBSD/i386 against Qt 5.15 (while 0.20.2 apparently did).

The same for i586, failing with a `terminate called after throwing an instance of 'std::bad_alloc'`. I wouldn't be surprised if this is due to a memory allocation of more than 2^31-1 (just over two GB of memory) being cast to an int32_t and therefor causing an attempt to allocate of a negative amount of memory. It consistently fails on the same file, which indeed still worked in 0.20.2.

**Comment by arjendekorte** [comment]
I've disabled the translations for i586 to make it build again with Qt 5.15.1. If the memory consumption keeps on increasing at this rate, I'm afraid we'll have to disable translations for all architectures in the not so distant future.

**Comment by danfe** [comment]
> I have no idea what lconvert is doing internally with GBs of memory.

It's actually not that bad.  The huge allocation happens in this loop of the function `loadPO()` inside `qttools-everywhere-src-5.15.0/src/linguist/shared/po.cpp`:
```c++
    QList<QByteArray> lines;
    while (!dev.atEnd())
        lines.append(dev.readLine().trimmed());
```
If I insert the `squeeze()` call in there, like this:
```c++
    QList<QByteArray> lines;
    while (!dev.atEnd()) {
        QByteArray line = dev.readLine().trimmed();
        line.squeeze();
        lines.append(line);
    }
```
or do `lines.append(line.data())` instead which apparently also causes release of unused memory, while proessing e.g. `stellarium-planetary-features/gu.po` file, `lconvert` allocates 63MB after reading all lines from file instead of 4026MB.

It would be nice if someone(tm) could review this and report the problem to the Qt developers.

**Comment by gzotti** [comment]
Thank you @danfe ! Yes, 63MB should be more like it. Alexander has reported this to Qt. Now let's see ...

**Comment by arjendekorte** [comment]
With the above patch, parallel builds are possible again: maximum memory usage is well below 1 GB on a 4 core build and `lconvert` doesn't seem to be the most memory intensive part of the build anymore

**Comment by gzotti** [comment]
See https://bugreports.qt.io/browse/QTBUG-87010: Has been solved - Thank you all! Qt5.15.2 should then work nicely again :-)

**Comment by arjendekorte** [comment]
Unfortunately, this patch didn't make it into Qt 5.15.2 https://github.com/qt/qtbase/blob/v5.15.2/src/corelib/io/qiodevice.cpp#L1487 so we're still stuck with running with in single thread

**Comment by gzotti** [comment]
I have also seen that Qt5.15.2 still seems to allocate all available RAM during translation, be it 4 or 12 GB, but I saw no failures. However I can build on a measly 4GB netbook with 2 cores. It just takes 4 hours. On an Core i7-2xxx notebook, it shows a zigzag pattern of RAM consumption (maxing out the 12GB, and possibly allocating virtual RAM, before falling back to about 6GB) during translation. C++ Compilation (MSVC) is totally harmless in comparison. This is something to report insist having fixed by Qt, and then releasing an update Qt5.15.3 still in the LTS/open-source area. We can do nothing, I am afraid.

**Comment by danfe** [comment]
I was also surprised to see that the fix did not appear in 5.15.2, because QTBUG-87010 references commits to both qt/qtbase/dev and qt/qtbase/5.15 branches.

**Comment by arjendekorte** [comment]
With all the excess memory being allocated, the builds are **really** slow. The QTBUG-87010 patch is backported in openSUSE Tumbleweed, so we'll be good there. The request to include this in Factory has been submitted already, so it should be available in a couple of days.
